### DISCUSSION TITLE
Machinery technical debt cleanup

### DISCUSSION THREAD

**Discussion by AffectedArc07** [ParadiseSS13/Paradise#18493] [Projects]
Right so, before the days of NanoUI and TGUI, we had a system where mobs held a ref to the machine you're currently interacting with. This is then what would have `Interact` called on it every so often to update the UI you are looking at.

As you can already imagine, this system is old and **awful**, for the following reasons:
- Ref holding to machines on mobs that will be a pain to track down
- We have TGUI and had NanoUI before which handles this in a much better way
- Its extra code to set and unset your active machine
- You know how when you try to change the pressure on those atmos consoles and it keeps tabbing you over? **THIS IS WHY**
- Definitely more
- GOOD GOD ITS A WASTE OF CPU

Let me elaborate on that last point

<details>
<summary>Technical ramble</summary>
This is the tick usage of gas canisters before I optimised them. 

![image](https://user-images.githubusercontent.com/25063394/180223221-e2b63a3a-e26f-482a-a1b5-4ea9b9d1a1c3.png)

The `updateDialog` proc isn't actually defined on the canister, and it has no "work" associated with it, its just a call to natives to get hearers and other mobs in range to decide who needs to update. Its horrible and uses 570us.

That sounds like a tiny amount right? Less than 1ms? Pathetic right? Wrong. Let me put that into context.

Our server runs at 20TPS, meaning we can easily get a tick duration with 1000 / 20 to get 50ms for the tick duration. However, that isn't the entire story. This tick has to handle running our code, as well as sending map updates to clients. Given that in highpop, maptick can reach upwards of 50-60% of the entire tick just to send map updates to players, the amount of time we have to do actual work gets a lot lower.

If we assume 50% of the tick is being used on map sending, we can then halve our 50ms to get 25ms of time for actual processing. That 570us of wasted CPU now accounts to 2.28% of the entire tick, which still doesn't sound like much, until you realise that this tick also has to handle the rest of atmospherics, machinery, player input, mob breathing, background database checks, everything. 

Wasting a full 2% of the tick on absolutely nothing is awful, and needs to be handled
</details>

---

List of machines that need refactoring to use TGUI to kill this inane system off:
- [ ] `/obj/machinery/abductor/console`
- [ ] `/obj/machinery/abductor/gland_dispenser`
- [ ] `/obj/machinery/abductor/experiment`
- [ ] `/obj/machinery/computer/cryopod` (#19540) 
- [ ] `/obj/machinery/disco`
- [ ] `/obj/machinery/hologram/holopad` - Make this just just an `alert()` selector for `Request AI` or `Holocall`
- [ ] `/obj/machinery/magnetic_controller` - AA has plans for this
- [ ] `/obj/machinery/navbeacon` - These have UIs?!
- [ ] `/obj/machinery/gameboard`
- [ ] `/obj/machinery/space_heater`
- [ ] `/obj/machinery/computer/arcade/battle`
- [ ] `/obj/machinery/computer/general_air_control` - AA IS WORKING ON THIS
- [x] `/obj/machinery/computer/prisoner` (#18505)
- [ ] `/obj/machinery/computer/merch`
- [x] `/obj/machinery/pipedispenser` - Can probably remove this now that we have RPDs (#18763)
- [ ] `/obj/machinery/implantchair`
- [ ] `/obj/machinery/arcade` - TGUIing this might be hard
- [ ] `/obj/machinery/arcade/claw` - OH GOD
- [ ] `/obj/machinery/computer/area_atmos`
- [ ] `/obj/machinery/bottler`
- [ ] `/obj/machinery/icemachine`
- [x] `/obj/machinery/juicer` - WHY DOES NONE OF THE KITCHEN STUFF HAVE A PROPER UI (#18764)
- [ ] `/obj/machinery/kitchen_machine`
- [ ] `/obj/machinery/plantgenes`
- [ ] `/obj/machinery/mineral/processing_unit_console` - Smelting console
- [ ] `/obj/machinery/mineral/stacking_unit_console` - These can both be removed now that we have the ORM
- [ ] `/obj/machinery/mineral/mint`
- [ ] `/obj/machinery/computer/turbine_computer`
- [ ] `/obj/machinery/particle_accelerator`
- [ ] `/obj/machinery/computer/pandemic`
- [ ] `/obj/machinery/reagentgrinder`
- [ ] `/obj/machinery/computer/telescience`
- [ ] `/obj/machinery/computer/rdservercontrol` - Being done in the R&D rework
- [ ] `/obj/machinery/r_n_d/experimentor` - I am just gonna remove this in the R&D rework

List of misc stuff that uses this wack system:
*I should point out that the var on mobs is casted to `/obj/machine` so all these existing is already a sin
- [x] `/obj/item/voodoo` - Dont think this is necessary (#18500)
- [ ] `/obj/item/spellbook` - Same here
- [ ] `/obj/item/paicard` - TGUI required
- [ ] `/obj/item/stack` - Crafting needs TGUIing
- [x] `/obj/item/teleportation_scroll` - Why isnt this just an area popup (#18495)
- [x] `/obj/item/implantpad` - TGUI (#18556)
- [ ] `/obj/item/storage/secure` - TGUI
- [ ] `/obj/structure/engineeringcart` - Make this an `input()` selector for `Select item to remove`
- [ ] `/obj/structure/janitorialcart` - Ditto
- [ ] `/obj/structure/noticeboard` - TGUI with options to `Show` and `Remove` documents
- [ ] `/obj/item/assembly` - Timers / igniters / stuff. I am 99% sure this is only used by timers.
- [ ] `/obj/item/assembly/infra`
- [ ] `/obj/item/cardhand`
- [ ] `/obj/structure/ore_box`
- [ ] The mob strip menu (`show_inv()`)
- [ ] `/obj/item/newspaper`
- [ ] `/obj/structure/filingcabinet` - See further up for noticeboards. Alternatively why not make those subtypes of this?
- [ ] `/obj/item/gun/energy/temperature`

There are definitely some things I have missed here, but anyone who contributes to this list will get major GBP gains. I want this stuff out my codebase before 2022 ends.

**Comment by GDNgit** [comment]
Claimed voodoo doll through #18500

**Comment by datlo** [comment]
Just noting that the pipe dispenser is still used by the ancient cryo station ruins and I've used it before to restore the oxygen supply.

Would be an improvement still to replace with an RPD. Maybe an ancient variant with a downside to fit the theme? Limited amount of devices?

**Reply by AffectedArc07** [reply]
I would be fine with just giving them an RPD as is, no nerfing required.

**Comment by Sirryan2002** [comment]
claiming merch computer as part of econ rework
